### PR TITLE
LF: retire ValueVersion

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
use TransactionVersion instead.

This is part of #7788

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
There will be a further missing reference to `ValueVersion` when merging to the main branch (in the newly added microbenchmarks).

**Comment by stefanobaghino-da** [comment]
> There will be a further missing reference to `ValueVersion` when merging to the main branch (in the newly added microbenchmarks).

Precisely. 🙂 

```
[2,441 / 2,710] scala //ledger/participant-integration-api:participant-integration-api; 60s remote-cache, worker ... (16 actions running)
ERROR: /home/vsts/agent/_work/3/s/ledger/participant-state/kvutils/tools/BUILD.bazel:130:23: scala //ledger/participant-state/kvutils/tools:benchmark-codec_generator failed (Exit 1): scalac failed: error executing command 
  (cd /home/vsts/.cache/bazel/_bazel_vsts/d5c8b3f1185416e5207d53317d3a7876/execroot/com_github_digital_asset_daml && \
  exec env - \
  bazel-out/k8-opt-exec-FFB32019/bin/external/io_bazel_rules_scala/src/java/io/bazel/rulesscala/scalac/scalac @bazel-out/k8-opt/bin/ledger/participant-state/kvutils/tools/benchmark-codec_generator_scalac_worker_input)
Execution platform: @io_tweag_rules_nixpkgs//nixpkgs/platforms:host
ledger/participant-state/kvutils/tools/codec-benchmark/src/benchmark/scala/com/daml/lf/benchmark/package.scala:17: error: object ValueVersion is not a member of package com.daml.lf.value
import com.daml.lf.value.{Value, ValueVersion}
       ^
ledger/participant-state/kvutils/tools/codec-benchmark/src/benchmark/scala/com/daml/lf/benchmark/package.scala:92: error: not found: value encodeVersionedValue
    encodeVersionedValue(CidEncoder, value, ValueVersion.DevOutputVersions)
    ^
ledger/participant-state/kvutils/tools/codec-benchmark/src/benchmark/scala/com/daml/lf/benchmark/package.scala:92: error: not found: value ValueVersion
    encodeVersionedValue(CidEncoder, value, ValueVersion.DevOutputVersions)
                                            ^
ledger/participant-state/kvutils/tools/codec-benchmark/src/benchmark/scala/com/daml/lf/benchmark/package.scala:17: warning: Unused import
import com.daml.lf.value.{Value, ValueVersion}
                                 ^
one warning found
three errors found
```

File: ledger/participant-state/kvutils/tools/codec-benchmark/src/benchmark/scala/com/daml/lf/benchmark/package.scala [hunk 1]
```diff
@@ -89,6 +89,6 @@ package object benchmark {
     encodeTransaction(NidEncoder, CidEncoder, transaction)
 
   private def encode(value: DecodedValue): EncodeResult[EncodedValue] =
-    encodeVersionedValue(CidEncoder, value, ValueVersion.DevOutputVersions)
+    encodeVersionedValue(CidEncoder, TransactionVersion.VDev, value)
```
Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/tools/codec-benchmark/src/benchmark/scala/com/daml/lf/benchmark/package.scala:92]: Probably didn't need my input, but the changes to the benchmark looks good. 👍🏻

Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/tools/codec-benchmark/src/benchmark/scala/com/daml/lf/benchmark/package.scala:92]: Thanks a lot ;)

**Review by miklos-da — APPROVED** [review]

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala [hunk 2]
```diff
@@ -247,4 +245,75 @@ object TransactionBuilder {
...
+          case FrontStackCons(value, values) =>
+            value match {
+              // for things supported since version 1, we do not need to check
+              case ValueRecord(_, fs) => go(currentVersion, fs.map(v => v._2) ++: values)
+              case ValueVariant(_, _, arg) => go(currentVersion, arg +: values)
+              case ValueList(vs) => go(currentVersion, vs.toImmArray ++: values)
+              case ValueContractId(_) | ValueInt64(_) | ValueText(_) | ValueTimestamp(_) |
+                  ValueParty(_) | ValueBool(_) | ValueDate(_) | ValueUnit =>
+                go(currentVersion, values)
+              case ValueNumeric(_) =>
```
Comment (hurryabit) [inline daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:272]: Is there a reason why this is not part of the previous case?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionVersion.scala [hunk 3]
```diff
@@ -29,14 +29,21 @@ object TransactionVersion {
...
+        Left(s"Unsupported transaction version '$vs'")
+    }
 
   def assertFromString(vs: String): TransactionVersion =
     data.assertRight(fromString(vs))
 
-  val minVersion = Values.min
-  private[transaction] val minChoiceObservers = VDev
-  private[transaction] val minNodeVersion = VDev
+  val minVersion: TransactionVersion = Values.min
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionVersion.scala:41]: Shouldn't this line somehow change given that we abandon value versions?

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/value/ValueCoderSpec.scala [hunk 4]
```diff
@@ -204,81 +155,45 @@ class ValueCoderSpec
...
-            ValueCoder.decodeVersionedValue(ValueCoder.CidDecoder, protoWithUnsupportedVersion)
-
-          actual shouldEqual Left(DecodeError(s"Unsupported value version ${badVer.protoValue}"))
-      }
   }
 
-  def testRoundTrip(value: Value[ContractId]): Assertion = {
-    val recovered = ValueCoder.decodeValue(
+  def testRoundTrip(version: TransactionVersion, value: Value[ContractId]): Assertion = {
+    val recovered = ValueCoder decodeValue (
```
Comment (hurryabit) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/value/ValueCoderSpec.scala:170]: Oh no, why are we removing the dot here?

**Review by hurryabit — APPROVED** [review]
This looks very reasonable to me. I really like how things get simpler and simpler. Thanks a lot.

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:272]: historical reason. I fix that.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionVersion.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionVersion.scala:41]: I am not sure what you mean. 

note that `Values` has nothing to do with LF values, it is the sequences of all possible values the type `TransactionVersion`. Should I use a different name ?

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/value/ValueCoderSpec.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/value/ValueCoderSpec.scala:170]: I do not remember having changed that. I will put it back.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionVersion.scala (hunk 3 above)
Comment (hurryabit) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionVersion.scala:41]: I think I've seen this called `All` in other places. Maybe that?

**Pull request merged**